File: 双指针/java/633.平方数之和.java
/*
 * @lc app=leetcode.cn id=633 lang=java
 *
 * [633] 平方数之和
 */

// @lc code=start
class Solution {
    public boolean judgeSquareSum(int c) {
        int left, right, fisrtValue, secondValue;
        left = 0;
        right = (int)Math.sqrt(c);
        while (left <= right) {
            fisrtValue = left * left;
            secondValue = right * right;
            //这两个加起来会有可能溢出，所以最好使用减法
            if (c - secondValue < fisrtValue) {
                right--;
            }
            else if (c - secondValue > fisrtValue) {
                left++;
            }
            else {
                return true;
            }
        }
        return false;
    }
}
// @lc code=end

File: 双指针/python/633.平方数之和.py
import math
#
# @lc app=leetcode.cn id=633 lang=python3
#
# [633] 平方数之和
#

# @lc code=start
class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        left, right = 0, int(math.sqrt(c))
        while left <= right:
            firstValue = left * left
            secondValue = right * right
            # 加法可能会造成数据溢出
            if c - firstValue > secondValue:
                left = left + 1
            elif c - firstValue < secondValue:
                right = right - 1
            else:
                return True
        
        return False
# @lc code=end

